sled-agent/src/instance_manager.rs

```diff
@@ -138,43 +140,66 @@ impl InstanceManager {
-                    .log
-                    .new(o!("instance" => instance_id.to_string()));
-                instances.insert(
-                    instance_id,
-                    Instance::new(
-                        instance_log,
+            match instances.get_mut(&instance_id) {
+                Some(instance) if migrate.is_none() => {
+                    // Instance already exists and we're not performing a migration
+                    info!(&self.inner.log, "instance already exists");
+                    (instance.clone(), None)
+                }
+                _ => {
+                    // Instance does not exist or one does but we're performing
+                    // a intra-sled migration. Either way - create an instance
+                    info!(&self.inner.log, "new instance");
+                    let instance_log = self
+                        .inner
+                        .log
+                        .new(o!("instance" => instance_id.to_string()));
+                    let old_instance = instances.insert(
                         instance_id,
-                        self.inner.nic_id_allocator.clone(),
-                        initial_hardware,
-                        self.inner.vlan,
-                        self.inner.nexus_client.clone(),
-                    )?,
-                );
-                let instance = instances.get_mut(&instance_id).unwrap().clone();
-                let ticket =
-                    Some(InstanceTicket::new(instance_id, self.inner.clone()));
-                (instance, ticket)
+                        Instance::new(
+                            instance_log,
+                            instance_id,
+                            self.inner.nic_id_allocator.clone(),
+                            initial_hardware,
+                            self.inner.vlan,
+                            self.inner.nexus_client.clone(),
+                        )?,
+                    );
+                    if let Some(old_instance) = old_instance {
+                        // If we had a previous instance, we must be migrating
+                        assert!(migrate.is_some());
+                        // TODO: assert that old_instance.inner.propolis_id() == migrate.src_uuid
+
+                        // We forget the old instance because otherwise if it is the last
+                        // handle, the `InstanceTicket` it holds will also be dropped.
+                        // `InstanceTicket::drop` will try to remove the corresponding
+                        // instance from the instance manager. It does this based off the
+                        // instance's ID. Given that we just replaced said instance using
+                        // the same ID, that would inadvertantly remove our newly created
+                        // instance instead.
+                        // TODO: cleanup source instance properly
+                        std::mem::forget(old_instance);
```

Comment:
Sorry for backpedaling, but I think I'm seeing your perspective from https://github.com/oxidecomputer/omicron/pull/447#discussion_r758911660 here... maybe the "null ticket" wouldn't be so bad.

I have a *minor* beef with this call to `forget` - we intend to drop the ticket, but forgetting the entire instance may also ignore other `drop` functions that could need to execute. I don't think those exist *today*, but there isn't anything preventing them from existing.

I think that justifies moving it out of the old instance, rather than creating a new ticket.

Something like:

```rust

// Ensure that the instance inserted into the `instances` mapping has a tracking
// membership ticket, so it can be safely removed when it goes out-of-scope.
let ticket = {
  if let Some(old_instance) = old_instance {
    // If we migrate an instance within the sled, we should keep it alive in the
    // `instances` mapping. Although the Propolis ID changes, the user-visible
    // instance ID remains the same, and uses the same key.
    assert!(migrate.is_some());
    std::mem::replace(old_instance.running_state.ticket, InstanceTicket::null());
  } else {
    InstanceTicket::new(instance_id, self.inner.clone())
  }
};

```

Reply:
>I have a minor beef with this call to forget - we intend to drop the ticket, but forgetting the entire instance may also ignore other drop functions that could need to execute. I don't think those exist today, but there isn't anything preventing them from existing.

Well, long-term I don't plan to just forget here :P Probably need to keep it around in case we need to rollback from a recoverable failure or just for cleaning up once the actual migration is done.

Reply:
Bumping this for visibility - It looks like we're still calling `forget` here; wasn't sure if we wanted to patch prior to merging.

Reply:
Yea, this still needs to be addressed but I planned on doing it in a followup PR. It will change slightly as I plumb more the propolis-side migration work which relies on both instances being available.